pkg/prowloader/github/github.go

```diff
@@ -15,63 +16,106 @@ type prlocator struct {
 	org    string
 	repo   string
 	number int
-	sha    string
+}
+
+type prentry struct {
+	mergedAt *time.Time
+	sha      string
 }
 
 type Client struct {
-	ctx    context.Context
-	client *gh.Client
-	cache  map[prlocator]*time.Time
+	ctx     context.Context
+	cache   map[prlocator]*prentry
+	prFetch func(org, repo string, number int) (*gh.PullRequest, error)
 }
 
 func New(ctx context.Context) *Client {
 	client := &Client{
 		ctx:   ctx,
-		cache: make(map[prlocator]*time.Time),
+		cache: make(map[prlocator]*prentry),
 	}
 	token := os.Getenv("GITHUB_TOKEN")
 	if token == "" {
-		logrus.Infof("No GitHub token environment variable, checking git config")
+		log.Infof("No GitHub token environment variable, checking git config")
 		var err error
 		token, err = gitconfig.GithubToken()
 		if err != nil {
-			logrus.WithError(err).Warningf("unable to retrieve GitHub token from git config")
+			log.WithError(err).Warningf("unable to retrieve GitHub token from git config")
 		}
 	}
 
+	var ghc *gh.Client
+
 	if token != "" {
 		ts := oauth2.StaticTokenSource(
 			&oauth2.Token{
 				AccessToken: token,
 			},
 		)
 		tc := oauth2.NewClient(client.ctx, ts)
-		client.client = gh.NewClient(tc)
+		ghc = gh.NewClient(tc)
 	} else {
-		logrus.Warningf("using unathenticated GitHub client, requests will be rate-limited")
-		client.client = gh.NewClient(nil)
+		log.Warningf("using unathenticated GitHub client, requests will be rate-limited")
+		ghc = gh.NewClient(nil)
+	}
+
+	client.prFetch = func(org, repo string, number int) (*gh.PullRequest, error) {
+		pr, _, err := ghc.PullRequests.Get(client.ctx, org, repo, number)
+		return pr, err
 	}
 
 	return client
 }
 
-func (c *Client) GetPRMerged(org, repo string, number int, sha string) (*time.Time, error) {
-	prl := prlocator{org: org, repo: repo, number: number, sha: sha}
-	if val, ok := c.cache[prl]; ok {
-		return val, nil
+// GetPRSHAMerged returns the merge time for a PR/SHA combination. The caching is designed
+// to minimize queries to GitHub. We basically have to handle these cases:
+//   - the PR doesn't exist (cache as nil)
+//   - the PR is unmerged (cache as nil)
+//   - the PR is merged with a different SHA (cache with the merged sha, return nil)
+//   - the PR is merged with the same SHA (cache with the merged sha, return merged time)
+func (c *Client) GetPRSHAMerged(org, repo string, number int, sha string) (*time.Time, error) {
+	prl := prlocator{org: org, repo: repo, number: number}
+	if val, ok := c.cache[prl]; ok && val != nil && val.sha == sha {
+		// If it's in the cache, and sha matches, this sha was merged.
+		return val.mergedAt, nil
+	} else if ok {
+		// If it's in the cache, but this isn't the sha, the PR merged but not with this sha.
```

Comment:
This should also have the unmerged case, right? We set the cache to nil below at line 113. Also, if during the database loading, subsequent PR updates will not be fetched until next run. How often do we run the db load now?

Reply:
We run `--load-database` hourly, it's a one-shot thing, it runs and then exits -- the cache only lives that long.

I fixed the comment, thanks!